Use AutoDiffScalar gradients in mex RBM method calls

tkoolen opened the pull request:
Instead of using user gradients, use AutoDiffScalar gradients in all relevant RBM mex functions.

Note that the current implementation will try to call method instantiations with a fixed maximum size derivative vector (with a max size up to 73, the number of states of a quat-parameterized Atlas) if possible, before gracefully falling back to true dynamic size derivative vectors. See https://github.com/RobotLocomotion/drake/issues/1199 for the performance comparison that inspired this choice.

I made it so that TaylorVar inputs to the functions also call mex functions.

In the process:
* added some more functionality to mexify:
 * added mexTryToCallFunctions, a function to which you can pass a bunch of std::functions, which will be tried in order until one sticks in the sense that the Matlab arguments passed in can be converted to the correct types required by the function ~~without a MexToCPPConversionError being thrown~~. I'm using this to switch between the three current options of double, auto diff fixed max size, and auto diff dynamic max size
 * a small infrastructure improvement regarding the number of outputs of the function
 * mex conversions to general std::vectors and std::pairs
 * improved conversion to AutoDiffScalar
 * moved conversion functions to standardMexConversions.h and rigidBodyManipulatorMexConversions.h
 * added some documentation
* mexified contactConstraintsmex
* fixed some more 'scalar correctness' issues in RBM (places where we were wrongly assuming doubles)
* implemented a (necessarily imperfect) make_function function that works for all of our use cases so far, and use it to clean up rbmMexFunctions
* fixed an issue in the gradients of FixedAxisOneDoFJoint frictionTorque gradients that was present both in Matlab and in C++
* added a Matlab sparse matrix to Eigen sparse conversion

~~This isn't quite ready to go in yet, but I wanted to get test results and to get a discussion started:~~
Ready to go after review and build servers say OK (MSVC might have trouble again).

1. ~~the mexTryToCallFunctions approach looks like it may have too much overhead when gradient output is required. I'll do some more benchmarking to make sure. Any ideas on clean alternatives or ways to improve performance?~~ **rewrote mexTryToCallFunctions without try/catch; much less overhead.**
1. The current return type of AutoDiffScalar methods is a TaylorVar matrix. Is that a good idea, or should I convert to a series of matrices? **returning TaylorVar is quite expensive due to the mexCallMatlabSafe call. Should probably return a bunch of matrices instead, but in a future iteration. Not a major concern.**
1. ~~This currently increases compilation time quite a bit due to the two extra sets of explicit RBM method instantiations. We could just completely switch to AutoDiff gradients and get rid of the user gradients to offset this, but I didn't want to make this pull request too big before discussing. What do you think?~~ **got rid of user gradients**
1. ~~More of a note to self: I have to make sure that all of the gradient tests still make sense, since gradients obtained by requesting more output arguments for a function and TaylorVar gradients are now computed by calling the exact same code.~~ **seems to be OK. Tests either don't use mex for TaylorVar gradient checks, or they use numerical gradients.**

tkoolen commented:
Internal compiler error on MSVC.

tkoolen commented:
MSVC: `fatal error C1060: compiler is out of heap space` while compiling RBM... Might need to get rid of the user gradients or find a different solution before this can be accepted.

tkoolen commented:
Got rid of the user gradients in RigidBodyManipulator. There are still some lower level things that use GradientVar, which I will clean up later.

tkoolen commented:
Updated description above.

tkoolen commented:
Do we care about the testIK timeout on drake006?

RussTedrake commented:
i do care.  but it's a known (to me) issue, not the fault of your pull 
request

> Twan Koolen <mailto:notifications@github.com>
> October 27, 2015 at 1:59 PMvia Postbox 
> <https://www.postbox-inc.com/?utm_source=email&utm_medium=sumlink&utm_campaign=reach>
>
> Do we care about the testIK timeout on drake006?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/RobotLocomotion/drake/pull/1434#issuecomment-151587872>.
>

RussTedrake commented on drake/systems/plants/@RigidBodyManipulator/RigidBodyManipulator.m:
```diff
@@ -406,7 +406,7 @@
         if compute_gradient
           ind = find(abs(v)<coulomb_window');
-          dind = sign(v(ind))./coulomb_window(ind)' .* coulomb_friction(ind)';
+          dind = coulomb_friction(ind)' ./coulomb_window(ind)';
```
why did you drop the `sign(v(ind))`?

RussTedrake replied:
nevermind.  it does look like the correct gradient for the line above.

tkoolen replied:
yeah, this was that one place where the gradients turned out to be wrong in both C++ and Matlab.

RussTedrake commented on drake/systems/plants/@RigidBodyManipulator/doKinematics.m:
```diff
@@ -88,8 +88,17 @@
+  if ~isnumeric(q)
+    if isa(q, 'TaylorVar')
+      if options.compute_gradients
+        error('can only compute first order gradients using mex')
```
i don't like this error message.  i think you're trying to avoid someone asking to compute the gradients of a taylorvar input (as opposed to using TaylorVar to compute the gradients of the nominal input).  but it took me a few minutes to parse the logic and figure out what you meant.

tkoolen replied:
Handling all of the possible inputs in doKinematics while trying to be efficient took a while to get right; I don't like how complicated this logic is currently...

Yes, I am indeed trying to avoid that case. I'll make the message more descriptive.

RussTedrake commented on drake/systems/plants/@RigidBodyManipulator/doKinematics.m:
```diff
@@ -99,6 +108,20 @@
+    nq = length(q);
+    nv = length(v);
+    if isnumeric(q)
+      q = TaylorVar(q, {[eye(nq), zeros(nq, nv)]});
```
This seems potentially inefficient; but I guess it keeps the mex interface abstraction clean?

tkoolen replied:
Yeah, that's the idea currently. We could perhaps use a struct with fields f and df to go back and forth between Matlab and C++ instead. As noted in the pull request description, there is also a bit of overhead in creating TaylorVars from mex as returned objects.

RussTedrake commented on drake/systems/plants/@RigidBodyManipulator/doKinematics.m:
```diff
@@ -99,6 +108,20 @@
   else
     kinsol.mex_ptr = options.kinematics_cache_ptr_to_use;
```
are you respecting the `force_new_kinsol` option here?

tkoolen replied:
There is only 'new kinsol' at this point. I should get rid of
```
if ~isfield(options, 'force_new_kinsol'), options.force_new_kinsol = false; end
```
which is the only remaining occurrence of `force_new_kinsol` in drake.

RussTedrake commented on drake/systems/plants/@RigidBodyManipulator/getCOM.m:
```diff
@@ -29,13 +29,22 @@
+      nq = length(kinsol.q);
+      dJ = reshape(dJ, numel(J), []); % convert to format where cols correspond to variables
+      dJ = dJ(:, 1 : nq); % make sure only gradients w.r.t. q are returned
+      dJ = reshape(dJ, size(J, 1), []); % convert to strange second derivative output format
```
perhaps it's finally time to revisit that one...?

tkoolen replied:
As in change the second derivative format that is used in Matlab to match C++?

RussTedrake replied:
Yes.  Maybe not today.  But soon.  

RussTedrake commented on drake/systems/plants/RigidBodyManipulator.h:
```diff
@@ -356,68 +246,65 @@ class DLLEXPORT_RBM RigidBodyManipulator
 
     // convert rotation representation
-    auto qrot = rotmat2Representation(R, rotation_type);
+    auto qrot = rotmat2Representation(T.linear(), rotation_type);
```
you change so many details in these big pull requests; i'm doing my best to check but it's very hard.

tkoolen replied:
This particular one was to clean up the signature of `relativeTransform`. It should have always returned a Transform, but we needed gradients, so we needed it to return a GradientVar corresponding to a 4x4 matrix instead. The method signature of `relativeTransform` was going to change anyway to get rid of GradientVar, so I took the opportunity to do it right.

Believe it or not, I do really try to minimize the number of changes in these pull requests. This one just had a large number of method signature changes that couldn't really be done gradually, and require all method calls to be adapted.

RussTedrake commented on drake/util/mexify.h:
```diff
@@ -8,89 +8,194 @@
 
 /**
- * fromMex
+ * Functions that make it easy to create a mex file that calls a std::function.
```
i like it

RussTedrake commented:
@tkoolen -- it looks amazing.  thank you for doing all this!  
i think you still have one legitimate test failing.  and I worry about big changes in performance -- have you done the benchmarking?
otherwise, to my (now very tired) eyes, it looks good to go. :)

tkoolen commented:
I'll take a look at the failing test.

As for performance, I think I did a pretty decent job of assessing the performance of the C++ code, see https://github.com/RobotLocomotion/drake/issues/1199 and https://github.com/RobotLocomotion/drake/pull/1350. My assumption here has been that forwardKinJacobian, massMatrix and dynamicsBiasTerm are representative methods.

I also benchmarked the overhead of the mexTryToCallFunctions approach and deemed it acceptable (after moving away from the try/catch approach).

I'll do one more assessment by running `benchmarkRigidBodyManipulator.m` both before and after these changes to make sure that I didn't introduce too much overhead on the Matlab side of things.

Thank you for the review!

tkoolen commented:
Addressed comments.

Benchmark results (benchmarkRigidBodyManipulator.m, times in ms per call):

Scenario 1: a couple of Jacobians

 | before | after
--- | --- | ---
no gradients | 0.5642 | 0.6280
with gradients | 1.549 | 2.546

Scenario 2: manipulatorDynamics

 | before | after
--- | --- | ---
no gradients | 0.2329 | 0.2681
with gradients | 3.924 | 3.077

There's quite a bit of overhead constructing and eval'ing TaylorVars, so I think we can get the gradient times down significantly by just using structs with f and df fields as the mex interface. In my opinion, that can wait until a later pull request though.

forwardKinmex has become quite a bit slower it seems:

Scenario 1

 | before | after
--- | --- | ---
forwardKinJacobianmex | 0.824 | 0.895
forwardKinmex | 0.156 | 0.564
doKinematicsmex | 0.360 | 0.317

Instruments showed that much of the time difference in forwardKinmex and forwardKinJacobianmex is attributable to checking that we're dealing with the right Scalar type in `KinematicsCache<Scalar>', which is currently done by doing a string compare between the output of `typeid' and the name stored in the DrakeMexPointer. This could definitely be improved a lot.

RussTedrake commented:
That’s a pretty severe performance hit (especially for forwardKinmex).  How hard is it to address?  Should we wait for it?



On November 22, 2015 at 11:27:57 PM, Twan Koolen (notifications@github.com) wrote:

Addressed comments.

Benchmark results (benchmarkRigidBodyManipulator.m, times in ms per call):

Scenario 1: a couple of Jacobians

before	after
no gradients	0.5642	0.6280
with gradients	1.549	2.546
Scenario 2: manipulatorDynamics

before	after
no gradients	0.2329	0.2681
with gradients	3.924	3.077
There's quite a bit of overhead constructing and eval'ing TaylorVars, so I think we can get the gradient times down significantly by just using structs with f and df fields as the mex interface. In my opinion, that can wait until a later pull request though.

forwardKinmex has become quite a bit slower it seems:

Scenario 1

before	after
forwardKinJacobianmex	0.824	0.895
forwardKinmex	0.156	0.564
doKinematicsmex	0.360	0.317
Instruments showed that much of the time difference in forwardKinmex and forwardKinJacobianmex is attributable to checking that we're dealing with the right Scalar type in KinematicsCache<Scalar>', which is currently done by doing a string compare between the output oftypeid' and the name stored in the DrakeMexPointer. This could definitely be improved a lot.

—
Reply to this email directly or view it on GitHub.

tkoolen commented:
Yeah sure, I can fix it. My idea is to have an unique integer for each C++ type that we'll use in a DrakeMexPointer, and to store that integer as an additional field in DrakeMexPointer.

tkoolen commented:
Some optimizations; effects turned out to be quite minor. See updated numbers in the previous post.

I found out why forwardKinmex seemed so slow: the 'after' included calls to forwardKinmex with TaylorVar inputs; comparing apples to oranges... So discard that and just look at the first two tables.

I think the overhead is still ridiculous (without gradients, only 11% of forwardKinmex time is spent in RigidBodyManipulator::forwardKin, 32% for forwardKinJacobianmex, 35% for dynamicsBiasTermmex, 59% for massMatrixmex), but it turned out to be hard to get rid of. Specifically mxGetProperty calls are very expensive and needed to process the DrakeMexPointers.

For scenario 1, I think the old gradients were faster because I actually implemented something better than just taking gradients of every operation in the algorithm (pretty much the only case of that).

RussTedrake commented:
interesting.  not asking for this now, but it should still be possible to implement user gradients (by populating the autodiff structures manually) if we choose... correct?  Perhaps it's worth adding a comment now in the code that used to be more optimized, with a link to a revision where we could find that code if we want to improve it later?

tkoolen commented:
> interesting. not asking for this now, but it should still be possible to implement user gradients (by populating the autodiff structures manually) if we choose... correct?

Yeah, you can override the method for AutoDiffScalar inputs and directly set the derivatives if you want. I'll put it in the 'rbm cleanup list' issue.

avalenzu commented:
Wow! Congrats, Twan!

On Thu, Nov 26, 2015, 4:06 AM Russ Tedrake <notifications@github.com> wrote:

> Merged #1434 <https://github.com/RobotLocomotion/drake/pull/1434>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/RobotLocomotion/drake/pull/1434#event-475365664>.
>

tkoolen commented:
Thanks!
